### PR TITLE
Horti v2

### PR DESCRIPTION
**Pull request by dianabarsan**
# Horti v2

Every ddoc stores a version, this is:
- the tag name for tags and releases
- the branch name + a timestamp for branch builds
- dev placeholder + a timestamp for local builds
Besides being saved as attachments to the staging document, built ddocs are stored in the API package (same format as the staging ddoc, a json file with a `docs` array property).

## On every API startup

The installation is checked: the ddocs present in every database (both staged and not staged) are compared to the bundled ddocs where:
- all bundled ddocs should exist in the database and
- they should have the same version

Scenarios:
- no differences found -> API just starts
- differences found in the uploaded ddocs, but the bundled ddocs are staged (same comparison as above, staged ddocs exist and have the correct version) -> API complete step (described below)
- differences found in the uploaded ddocs, and at least some bundled ddocs are not staged -> fresh install (described below)

### Prep
The previous incomplete installation attempt, if found, is aborted. (in-memory stored upgrade_log id or latest non-final upgrade log). 

Each installation or upgrade should have its own upgrade_log - this is a document that is saved in the medic-logs DB (???) and has a structure like:

```json
{
  "_id": "upgrade_log:<start_timestamp>:<upgrade_to_version>",
  "user": "<initiating user name, for upgrades>",
  "action": "<install, stage or upgrade>",
  "from_version": "<current version, empty when doing fresh install>",
  "to_version": "<future version, bundled version when doing fresh install>",
  "start_date": "<timestamp>",
  "state": "<one of: initiated, staged, indexing, indexed, completing, complete, aborted, errored>",
  "updated_date": "<timestamp>",
  "state_history": [{ "state": "<state>", "date": "<timestamp>" }]
}
```

The _id of the upgrade log is saved in memory. If API restarts, it will search for the latest created upgrade log, that is not in a final state (finalized, errored or aborted). <- this part might need more work. 
The upgrade_log document (has state of "initiated").

Prep stage is not executed for partial installs (when api starts and some ddocs exist and some do not). 

### Stage
Existent uploaded staged ddocs are deleted. (view cleanup is NOT called)
For upgrades, the definitions of the ddocs are downloaded from the staging server. For fresh installs, the bundled definitions are used in the next steps.

The ddocs (bundled or downloaded) are renamed to add the ":staged:" prefix (_design/design_doc_name -> _design/:staged:design_doc_name), assigned partial `build_info` and uploaded to their corresponding databases. At this stage, the `build_info` consists of:
```json
{
  "_id": "_design/something",
  "build_info": {
    "user": "<initiating user from upgrade_log>",
    "upgrade_log_id": "<the _id of the upgrade_log>"
  }
}
```
The upgrade_log state is set to "staged". 

### View indexing

The upgrade_log state is set to "indexing".

Iterates over all uploaded staged ddocs, for every known database, making a list of requests that start view indexing. These requests have a retry mechanism when the failure is a timeout. 
Starts querying `active_tasks` on a timeout (5 seconds) and calls the view indexing requests. The "summary" of `active_tasks` is logged to api logs (identical to the indexer progress in horti). 
Compared to `horti`, the results of the indexing progress summary are _not_ saved in the upgrade_log document. (big spam, think about it?)
View indexing resolves when all view indexing requests get 2xx statuses. Querying `active_tasks` is stopped.

The upgrade_log state is set to "indexed".

### Completing the upgrade (rename this to something else? finalizing? executing?)

This is not part of this PR. This will consist on calling the "upgrader" service upgrade endpoint that is supposed to replace the source code for API and Sentinel.
Big TODO

### Completing by API startup 

As part of the "installation check" mentioned above, we will find that either: 
- (1) all ddocs (for all dbs) are staged 
- (2) some dbs lack staged ddocs

#### For (2)

When not all ddocs are staged, we re-start with "Stage", "View indexing" and then finally "Completing".
Because we initiate the installation code in the old API, which is not aware of the new database existing, those ddocs are ignored.
Because we do not do view_cleanup at any of these stages, the old staged ddocs indexes are preserved, and we only end up indexing view for "new" databases (which should be empty). 

#### For (1), actual completing

TODO: get the upgrade_log _id from any of the staged ddocs. 
The upgrade_log state is set to "completing".
All staged ddocs are renamed, to remove the :staging: prefix, the current timestamp is saved in their `deploy_info`, and then are saved, thus overwriting old ddocs. 
TODO: instead of renaming ALL staged ddocs, match them against the ddocs in the bundle, and only rename the matching ones, delete the others.
The upgrade_log state is set to "complete".
Run compaction and view_cleanup on all databases. 

If any of the steps fails, the upgrade_log doc gets the status of "errored". 

### GET /api/v2/upgrade (new endpoint)
- returns "current" upgrade_log contents and indexer progress
- will be queried by the admin app?
```json
{
  "upgradeDoc": { ... },
  "indexers": [{ "database", "ddoc", "progress" }],
}
```

### POST /api/v2/upgrade 
- needs version in payload
- starts an upgrade:
  - prep sync (request is resolved after prep is finished) 
  - stage async
  - view indexing async
  - complete async 

### POST /api/v2/upgrade/stage
- needs version in payload
- stages an upgrade:
    - prep sync (request is resolved after prep is finished)
    - stage async
    - view indexing async     

### POST /api/v2/upgrade/complete
- calls the complete step, which is TODO

### DELETE /api/v2/upgrade (new endpoint)
- aborts the current upgrade 
  - deletes staged ddocs
  - sets the upgrade_log state to "aborted"
  - deletes the upgrade_folder

This PR does not make changes to the Admin app (in the next PR), and does not add e2e tests for the upgrade process. 
Given that, with horti, we always end up with one ddoc (_design/medic), we're always in a "partial" upgrade state. Any e2e tests added at this point would need to work around that by adding some temporary code. I'd prefer to add final e2e tests, that would also cover partial upgrades, once we dockerize the services.

I think more "safety" measures, or edge case handling can be added throughout, also some things might need to get reworked as development progresses and we add the upgrade service. 

medic/cht-core#7406

# Code review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

**Comment by dianabarsan** [comment]
hey @garethbowen 

This is very much a work in progress, but it'd be great to get some feedback early before overcommitting. Could you please take a look? Thanks a lot!

File: scripts/build/index.js [hunk 1]
```diff
@@ -115,15 +116,30 @@ const updateServiceWorker = () => {
     }
 
     if (err.error && err.error.code === 'ECONNREFUSED') {
-      console.warn('API could not be reached, so the service-worker has not been updated.');
+      console.warn('API could not be reached.');
```
Comment (garethbowen) [inline scripts/build/index.js:119]: Regression?

File: api/src/controllers/upgrade.js [hunk 2]
```diff
@@ -1,11 +1,14 @@
 const auth = require('../auth');
 const serverUtils = require('../server-utils');
 
-const service = require('../services/upgrade');
+const service = require('../services/horti/upgrade');
```
Comment (garethbowen) [inline api/src/controllers/upgrade.js:4]: Can we drop the "horti" name altogether? Nobody gets the joke  :/

**Review by garethbowen — COMMENTED** [review]
It's all looking great so far.

I'm not quite sure about the way we're doing secrets. Firstly the name threw me off because it's not really secret it's just a unique ID for that specific build. Secondly, the way I see it we have three options for this: (a) a generated UUID (what you've done), (b) the semver version with an alpha component for development, (c) a hash of the ddoc.

I think (b) is better than (a) because it's human readable and already exists (on the main ddoc at least). It's also easier to find the source if someone tells you which version of the ddoc they're using rather than which UUID it has (eg: I can tell roughly how old 3.4.0 is but a UUID gives you no clue).

I think (c) is good if we are trying to verify the correctness of the ddoc, but we'd have to distribute the hash independently from the ddoc otherwise it's easy to also update the hash and bypass any validation we have.

Unless I've missed something... why are you generating a UUID instead of using semver?

File: scripts/build/index.js (hunk 1 above)
Comment (dianabarsan) [reply scripts/build/index.js:119]: Bad merge :)

File: api/src/controllers/upgrade.js (hunk 2 above)
Comment (dianabarsan) [reply api/src/controllers/upgrade.js:4]: Yea haha definitely.

**Comment by dianabarsan** [comment]
hey @garethbowen 

I added a detailed overview of what this PR contains. I'd be happy to talk through it and discuss possibilities or changes or things I missed or future plans. 
The code is also available for review. 

Thanks a lot for your time! Please let me know if you have an idea about how to make this process more efficient.

File: api/src/services/setup/upgrade.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+      return;
+    }
+    await complete();
+  } catch (err) {
+    await upgradeLogService.setErrored();
+    logger.error('Error thrown when indexing views %o', err);
+  }
+};
+
+const complete = async () => {
```
Comment (garethbowen) [inline api/src/services/setup/upgrade.js:34]: Can we change this to `finalize`, and maybe update the associated state to `finalizing`? `complete` can be interpreted either as it _is_ complete already, or a command to complete it.

**Review by garethbowen — CHANGES_REQUESTED** [review]
Great work!

After reading your excellent wall of text, I skimmed the code and left some minor comments inline. On top of those I have two broad comments...

1. I don't quite understand why we need to copy the upgrade log from the db to the filesystem, and it seems risky if the two get out of sync.
2. You've broken backwards compatibility with at least some of the upgrade APIs so I think we should bump it to v2 while keeping v1 around and reworking the request body to maintain backwards compatibility.

Happy to talk these over in a call or on Slack - let me know!

File: api/src/services/setup/upgrade-log.js [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+  // temporary patching to work on Node 8.
+  // This code will never run on Node 8 in prod!
+  fs.promises = {
+    mkdir: promisify(fs.mkdir),
+    readdir: promisify(fs.readdir),
+    rmdir: promisify(fs.readdir),
+    unlink: promisify(fs.unlink),
+    access: promisify(fs.access),
+    writeFile: promisify(fs.writeFile),
+    readFile: promisify(fs.readFile),
```
[the remark this answers is not in the document]
Comment (dianabarsan) [reply api/src/services/setup/upgrade-log.js:34]: Yes, I know. These are just hacks so I get e2e tests to pass on Node 8. That's in the next PR :D

File: api/src/services/setup/upgrade.js (hunk 3 above)
Comment (dianabarsan) [reply api/src/services/setup/upgrade.js:34]: yes!

File: api/src/controllers/upgrade.js [hunk 5]
```diff
@@ -1,40 +1,62 @@
 const auth = require('../auth');
 const serverUtils = require('../server-utils');
 
-const service = require('../services/upgrade');
+const service = require('../services/setup/upgrade');
 const configWatcher = require('../services/config-watcher');
 
+const REQUIRED_PERMISSIONS = ['can_configure'];
```
[the remark this answers is not in the document]
Comment (dianabarsan) [reply api/src/controllers/upgrade.js:7]: I think there already is an issue scheduled for `3.15`: https://github.com/medic/cht-core/issues/7139
Do you think we should move that to 4.0?

**Comment by dianabarsan** [comment]
> I don't quite understand why we need to copy the upgrade log from the db to the filesystem, and it seems risky if the two get out of sync.

I had no better way to tracking the "current" upgrade doc. 
At first, I just saved the _id of the doc in the file system. When actually needing the contents, I always `get` it from the database. 

Do you think it would be clearer if : 
a) I just saved the _id of the document in the FS, instead of the whole contents
b) I changed the name to be `upgrade_log:<timestamp>:<version>` and use _all_docs to get the latest one? 
c) Add a view to get the log by date and keep the same _id format?

Thanks a lot for taking the time!

File: api/src/controllers/upgrade.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (garethbowen) [reply api/src/controllers/upgrade.js:7]: Ahh that makes sense. Let's leave it in 3.15 for now. It could be considered a breaking change because users will lose access to that page, but I suspect everyone upgrading is a full admin anyway...

**Comment by garethbowen** [comment]
I'd like to avoid the FS altogether if possible to avoid any inconsistencies. It could cause problems if the api container had to be thrown away and recreated if something goes wrong. Furthermore if we end up wanting to run multiple api containers for failover or scalability reasons it would cause issues.

I like your option (b) though do we ever query by version?

I worry that (c) would use a fair bit of processing time going over all the other log docs, but if we need to query by version AND timestamp then it may be necessary.

**Comment by dianabarsan** [comment]
> do we ever query by version?

Convenience if someone wants to quickly get the upgrade_log for upgrade to a specific version. 
I'll go with the changed name, then! Thanks

**Comment by dianabarsan** [comment]
Hey Gareth!

I pushed a bunch of changes: 
1. dropped fs for storing the log
2. instead storing the "current" upgrade id in memory and re-read (optimistically) from the DB if api restarts. I need to test this more. 
3. Added v2/upgrade with the new endpoints
4. moved some code around. 

I updated the description above to describe the changes.

**Comment by dianabarsan** [comment]
Feedback welcome! 
Thank you!!

File: api/src/services/setup/upgrade-log.js [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+};
+
+const UPGRADE_LOG_NAME = 'upgrade_log';
+
+const getUpgradeLogId = (version, startDate) => `${UPGRADE_LOG_NAME}:${startDate}:${version}`;
+
+const getLatestUpgradeLog = async () => {
+  const now = new Date().getTime();
+  const results = await db.medicLogs.allDocs({
+    startkey: `${UPGRADE_LOG_NAME}:${now}:`,
```
Comment (garethbowen) [inline api/src/services/setup/upgrade-log.js:47]: Adding `${now}` doesn't change anything as far as I can tell, because we shouldn't have any logs from the future. Can we just have...

```suggestion
    startkey: `${UPGRADE_LOG_NAME}:`,
```

File: api/src/services/setup/upgrade-log.js [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+
+const setCompleting = async () => {
+  logger.info('Completing install');
+  await update(UPGRADE_LOG_STATES.COMPLETING);
+};
+
+const setComplete = async () => {
+  await update(UPGRADE_LOG_STATES.COMPLETE);
+  logger.info('Install complete');
+};
```
Comment (garethbowen) [inline api/src/services/setup/upgrade-log.js:199]: It looks like these functions are not called anywhere - are they just left over from the `finalize` rename?

**Review by garethbowen — COMMENTED** [review]
Awesome work. Really happy with this approach.

I've left a couple of minor comments.

Let me know if it's ready for a full review.

File: api/src/services/setup/upgrade-log.js (hunk 7 above)
Comment (dianabarsan) [reply api/src/services/setup/upgrade-log.js:199]: They're going to get called in the "complete" endpoint of the API (where the services are restarted). 
That part is not developed yet.

File: api/src/services/setup/upgrade-log.js (hunk 6 above)
Comment (dianabarsan) [reply api/src/services/setup/upgrade-log.js:47]: It's just so somebody manually creating a upgrade_log from 2032 will screw up all of our upgrade attempts :D 
I'd rather keep it.

**Comment by dianabarsan** [comment]
> Let me know if it's ready for a full review.

I think that, with the off time that I'll have, it can go through a full review so we don't delay any longer, and I'll have feedback to work on when I get back. How does that sound?

**Comment by garethbowen** [comment]
Perfect. It'll be sitting on your desk when you return!

**Review by garethbowen — CHANGES_REQUESTED** [review]
Excellent work. Just the usual minor improvements and nitpicks.

File: api/src/services/setup/upgrade-log.js [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+  INDEXED: 'indexed',
+  COMPLETING: 'completing',
+  FINALIZING: 'finalizing',
+  FINALIZED: 'finalized',
+  COMPLETE: 'complete',
+  ABORTED: 'aborted',
+  ERRORED: 'errored',
+};
+
+let currentUpgradeLogId;
```
[the remark this answers is not in the document]
Comment (dianabarsan) [reply api/src/services/setup/upgrade-log.js:32]: > I still have concerns about caching this. Is getting the most recent upgrade log from the DB too expensive, or potentially incorrect?

Because it is potentially incorrect. There's also less potential for race conditions, if two users click on the upgrade button at the same time. 

> if we ever decide to run multiple API containers

This is why I used FS before. Would be super nice to have a proper caching service :)

File: api/src/services/setup/utils.js [hunk 9]
```diff
@@ -0,0 +1,276 @@
...
+const setStagingData = (ddocs, deployInfo) => {
+  if (!ddocs || !ddocs.length) {
+    return [];
+  }
+
+  return ddocs.map(ddoc => {
+    ddoc._id = ddocsService.stageId(ddoc._id);
+    ddoc.deploy_info = deployInfo;
+    return ddoc;
+  });
```
[the remark this answers is not in the document]
Comment (dianabarsan) [reply api/src/services/setup/utils.js:95]: > How is this working? If you change the _id shouldn't you also update the _rev?

These docs are coming from either: 
1. The "local" store of bundled ddocs (and shouldn't have a rev)
2. The "downloaded" bundle of the new version (and shouldn't have a rev). 
They aren't read from the database. I'll add a delete _rev just to be safe.

File: api/src/services/setup/view-indexer-progress.js [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+  });
+};
+
+const calculateAverageProgress = (indexer) => {
+  const tasks = Object.keys(indexer.tasks);
+  indexer.progress = Math.round(tasks.reduce((progress, pid) => progress + indexer.tasks[pid], 0) / tasks.length);
+};
+
+// logs indexer progress in the console
+// _design/doc  [||||||||||29%||||||||||_________________________________________________________]
```
[the remark this answers is not in the document]
Comment (dianabarsan) [reply api/src/services/setup/view-indexer-progress.js:62]: This already existed in the horti code: https://github.com/medic/horticulturalist/blob/master/src/install/warmViews.js#L25
I just copied it.

File: api/src/services/setup/upgrade-log.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (dianabarsan) [reply api/src/services/setup/upgrade-log.js:32]: I still believe that the local cache saves us from potential race conditions and incorrect DB lookup results. 

Do you think we can bank on having a caching service when we decide to add multiple APIs? We will already need it, because of other things we cache.

File: api/src/controllers/upgrade.js [hunk 11]
```diff
@@ -1,40 +1,78 @@
...
     })
+    .then(() => res.json({ ok: true }))
+    .catch(err => serverUtils.error(err, req, res));
+};
+
+const completeUpgrade = (req, res) => {
+  return checkAuth(req)
+    .then(() => service.complete().then(() => res.json({ ok: true })))
+    .catch(err => serverUtils.error(err, req, res));
+};
```
[the remark this answers is not in the document]
Comment (dianabarsan) [reply api/src/controllers/upgrade.js:44]: Cool, I like this idea!
I'm adding code to pass the payload of the complete call to the service. The service function isn't implemented yet, but I added a comment to implement the validation.

File: api/src/routing.js [hunk 12]
```diff
@@ -336,10 +337,22 @@ app.get('/api/info', function(req, res) {
   res.json({ version: p.version });
 });
 
-app.get('/api/deploy-info', (req, res) => {
+app.get('/api/deploy-info', async (req, res) => {
   if (!req.userCtx) {
     return serverUtils.notLoggedIn(req, res);
   }
+
+  // todo move this to some service or something
```
[the remark this answers is not in the document]
Comment (dianabarsan) [reply api/src/routing.js:345]: If I add it to `environment`, it will create a circular dependency, where `db` requires `environment` and `environment` requires `db`. This is the only reason it isn't there and this silly todo comment. 
Do you think I should just require `db` at runtime, in the `environment` function?

File: api/src/services/setup/view-indexer.js [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+
+  do {
+    try {
+      await rpn.get({
+        uri: `${environment.serverUrl}/${dbName}/${ddocId}/_view/${viewName}`,
+        json: true,
+        qs: { limit: 1 },
+        timeout: 2000,
+      });
+      viewIndexed = true;
```
[the remark this answers is not in the document]
Comment (dianabarsan) [reply api/src/services/setup/view-indexer.js:68]: Yea! Funny enough, there's an eslint rule [`no-constant-condition`](https://eslint.org/docs/rules/no-constant-condition) which complains when you do `while(true)`. So a variable is still needed to satisfy it, or ignore the rule.

File: api/tests/mocha/controllers/upgrade.js [hunk 14]
```diff
@@ -125,10 +162,115 @@ describe('Upgrade controller', () => {
...
+      sinon.stub(service, 'indexerProgress').resolves(['indexers']);
+
+      await controller.upgradeInProgress(req, res);
+
+      service.upgradeInProgress.callCount.should.equal(1);
+      service.indexerProgress.callCount.should.equal(1);
+      res.json.callCount.should.equal(0);
+      serverUtils.error.callCount.should.equal(1);
+    });
+
```
[the remark this answers is not in the document]
Comment (dianabarsan) [reply api/tests/mocha/controllers/upgrade.js:226]: Ah oops, removed.

**Comment by dianabarsan** [comment]
I pushed a bunch of small changes: 
- Added a new service to get/set deploy-info
- Changed the payload of POST v2/upgrade to match the POST v1/upgrade (ie passing the build_info instead of just the version). I still kept v2 and moved all the new endpoints there. 
- Fixed a bug where aborting an upgrade did not stop view indexer query (the more I add to this, the more I think we should have some upgrade "state" saved in memory to make sure processes can communicate and react to input). 
- Changed a check that would have always used the local ddocs when the upgrade to the same version was requested, so it will download the docs from staging every time. I'm still debating what to do on this one, the use-case I'm thinking of is a QA testing and wating to install the latest build of the branch that they are already running. I think downloading the docs from staging is more correct anyway. The "local" docs are only used when checking the installation, on boot. 
- Addressed he PR feedback

File: api/src/routing.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (garethbowen) [reply api/src/routing.js]: D'oh! No, I'm not a fan of runtime requires. Leave it as is or make a separate service, all good.

File: api/src/services/setup/upgrade-log.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (garethbowen) [reply api/src/services/setup/upgrade-log.js:32]: > This is why I used FS before.

I don't think FS saves us in any way, unless the two containers share a common mount point, right?

> Because it is potentially incorrect. There's also less potential for race conditions, if two users click on the upgrade button at the same time.

I don't follow. With the single container, If two users click go at the same time then two docs will be created and the one that's slightly later will overwrite the `currentUpgradeLogId` in memory right? You'd have two upgrades initiated but only one doc actually being used, and both users would get info from that doc. If instead we relied on the latest doc in the DB, then both users would write a separate doc, the code would only use the latest one, and you'd have the exact same scenario - the slightly later one would be used and the first one would be orphaned.

To be clear, what I'm proposing is instead of storing the ID of the latest doc, we just query the DB for the latest doc. The only benefit I can see with storing it is slightly improved performance (ID lookup rather than view query). The big downside to storing the ID is if some coder forgets to invalidate the cache it'll behave unpredictably.

File: api/src/services/setup/view-indexer.js [hunk 15]
```diff
@@ -55,23 +55,21 @@ const getViewsToIndex = async () => {
  * @return {Promise}
  */
 const indexView = async (dbName, ddocId, viewName) => {
-  let viewIndexed = false;
-
+  const indexing = true;
```
Comment (garethbowen) [inline api/src/services/setup/view-indexer.js:58]: I don't think this is necessary... just use `while(true)`!

**Review by garethbowen — CHANGES_REQUESTED** [review]
Comments inline. Help me understand the race condition concern.

File: api/src/services/setup/utils.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (garethbowen) [reply api/src/services/setup/utils.js]: Ahh cool, cool. Thanks for the explanation!

File: api/src/services/setup/view-indexer.js (hunk 15 above)
Comment (dianabarsan) [reply api/src/services/setup/view-indexer.js:58]: Eslint complains :D 

> Yea! Funny enough, there's an eslint rule [no-constant-condition](https://eslint.org/docs/rules/no-constant-condition) which complains when you do while(true). So a variable is still needed to satisfy it, or ignore the rule.

https://github.com/medic/cht-core/pull/7435#discussion_r805928558

I preferred the variable to eslint ignore.

File: api/src/services/setup/upgrade-log.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (dianabarsan) [reply api/src/services/setup/upgrade-log.js:32]: I think it's slightly more correct to keep an upgrade "state" in memory. 
I already have two other places that need it: 
1. the view progress query - that needs to stop on cancel
2. the view indexing queries - that will throw an error if I delete the stated view from under it's feet

For 1, I already added a workaround - by also keeping a reference to the "stop" function in memory. For 2, I switched two calls so the upgrade_log gets the "abort" state before the "errored" state, but the error still gets logged and I'd prefer to catch it. I added a TODO there. 

I'm even inclining to have `prep` immediately create a state object with the necessary bits that gets passed around, and all pending processes can be gracefully handled on a stop. 

Please tell me what you think.

File: api/src/routing.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (dianabarsan) [reply api/src/routing.js]: I had already added a separate service :)

File: api/src/services/setup/view-indexer.js (hunk 15 above)
Comment (garethbowen) [reply api/src/services/setup/view-indexer.js:58]: What a BS rule... apparently `while(true)` is bad but `for (;;) {}` is ok?!

Comment (garethbowen) [reply api/src/services/setup/view-indexer.js:58]: It makes sense for if/else... maybe we should set `checkLoops:false`?

File: api/tests/mocha/services/setup/view-indexer.spec.js [hunk 16]
```diff
@@ -0,0 +1,219 @@
...
+
+  describe('stopIndexing', () => {
+    it('should set flag to stop indexing', () => {
+      viewIndexer.__set__('continueIndexing', true);
+      viewIndexer.stopIndexing();
+      expect(viewIndexer.__get__('continueIndexing')).to.equal(false);
+
+      viewIndexer.stopIndexing();
+      expect(viewIndexer.__get__('continueIndexing')).to.equal(false);
+    });
```
Comment (garethbowen) [inline api/tests/mocha/services/setup/view-indexer.spec.js:164]: This is probably unnecessary - it's just testing a setter, right?

I think there's a more version of this test which instead of just checking the internal state, checks that the indexing loop stops after calling `stopIndexing()`.

**Review by garethbowen — APPROVED** [review]
Love it. Great work!

I've left a comment in about a slight improvement to a unit test, but I'll leave that up to you.
